src/matrix-to.js

```diff
@@ -14,21 +14,107 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+import MatrixClientPeg from "./MatrixClientPeg";
+
 export const host = "matrix.to";
 export const baseUrl = `https://${host}`;
 
+// The maximum number of servers to pick when working out which servers
+// to add to permalinks. The servers are appended as ?via=example.org
+const MAX_SERVER_CANDIDATES = 3;
+
 export function makeEventPermalink(roomId, eventId) {
-    return `${baseUrl}/#/${roomId}/${eventId}`;
+    const serverCandidates = pickServerCandidates(roomId);
+    return `${baseUrl}/#/${roomId}/${eventId}?${encodeServerCandidates(serverCandidates)}`;
 }
 
 export function makeUserPermalink(userId) {
     return `${baseUrl}/#/${userId}`;
 }
 
 export function makeRoomPermalink(roomId) {
-    return `${baseUrl}/#/${roomId}`;
+    const serverCandidates = pickServerCandidates(roomId);
+    return `${baseUrl}/#/${roomId}?${encodeServerCandidates(serverCandidates)}`;
 }
 
 export function makeGroupPermalink(groupId) {
     return `${baseUrl}/#/${groupId}`;
 }
+
+export function encodeServerCandidates(candidates) {
+    if (!candidates) return '';
+    return `via=${candidates.map(c => encodeURIComponent(c)).join("&via=")}`;
+}
+
+export function pickServerCandidates(roomId) {
+    const client = MatrixClientPeg.get();
+    const room = client.getRoom(roomId);
+    if (!room) return [];
+
+    // Permalinks can have servers appended to them so that the user
+    // receiving them can have a fighting chance at joining the room.
+    // These servers are called "candidates" at this point because
+    // it is unclear whether they are going to be useful to actually
+    // join in the future.
+    //
+    // We pick 3 servers based on the following criteria:
```

Comment:
Shouldn't we prioritise suggesting the server where the room was created?

Reply:
or is the idea that the receiving server should infer a top priority candidate from the server_name in the room_id?

Reply:
There's not a whole lot of guarantee that `example.org` in `!somewhere:example.org` is still a resident of the room. Ideally Synapse would use the 3 we suggest *and* the server in the room ID as a last ditch attempt.

I'm failing to find the issue for this though